File: src/Analysis.cpp
#include "Analysis.hpp"
#include <queue>
#include <algorithm>
#include <iostream>

Analysis::Analysis(Graph& g, Config& c) : graph(g), config(c) {
    library = config.getLibrary();
}

std::vector<AnalysisResult> Analysis::run(double& worstSlack, std::string& worstNode) {
    propagateArrivalTimes();
    calculateRequiredTimes();
    return calculateSlack(worstSlack, worstNode);
}

void Analysis::propagateArrivalTimes() {
    std::cout << "Propagating Arrival Times..." << std::endl;
    auto nodes = graph.getAllNodes();
    
    // Reset AT
    for (Node* n : nodes) n->at = -1.0;

    std::queue<Node*> q;
    std::map<Node*, int> inDegree;

    // Calculate in-degree for topo sort
    // Only count edges that matter for propagation? Graph includes internal and net edges.
    for (Node* n : nodes) inDegree[n] = 0;
    
    for (Node* n : nodes) {
        for (auto& edge : n->edges) {
            inDegree[edge.target]++;
        }
    }

    // Identify start points and seed queue
    for (Node* n : nodes) {
        bool isStartPoint = false;
        
        // DFF Outputs (Q) - Simple heuristic matching Python
        if (n->name.find("/Q") != std::string::npos && n->name.find("reg_") != std::string::npos) {
            n->at = library["cells"]["DFF"]["delay_clk_q"];
            isStartPoint = true;
        }
        
        // Input Ports (checking if type is "port" and no incoming edges or explicitly named)
        // Since my parser sets type="port" for inputs, let's use that.
        // Also check if node is "rst" or "data_in" based on Python logic?
        if (n->type == "port" && inDegree[n] == 0) {
             n->at = config.getInputDelay();
             isStartPoint = true;
        }

        if (isStartPoint) {
            // If it's a start point, it should have initial AT.
            // But for Topo sort, we start with nodes having in-degree 0.
            // Wait, DFF/Q is *internally* a start point but it might not have in-degree 0 in the full graph?
            // Actually DFF/Q has NO incoming edges in our graph because we break loops at DFF!
            // So DFF/Q should have inDegree 0.
        }
    }

    // Topological Sort Queue
    for (Node* n : nodes) {
        if (inDegree[n] == 0) {
            q.push(n);
        }
    }

    std::vector<Node*> topoOrder;
    while (!q.empty()) {
        Node* u = q.front();
        q.pop();
        topoOrder.push_back(u);

        for (auto& edge : u->edges) {
            Node* v = edge.target;
            inDegree[v]--;
            if (inDegree[v] == 0) {
                q.push(v);
            }
        }
    }

    // Propagate
    for (Node* u : topoOrder) {
        if (u->at == -1.0) continue;

        for (auto& edge : u->edges) {
            Node* v = edge.target;
            double newAt = u->at + edge.delay;
            if (newAt > v->at) {
                v->at = newAt;
            }
        }
    }
}

void Analysis::calculateRequiredTimes() {
    std::cout << "Calculating Required Times..." << std::endl;
    double period = config.getClockPeriod();
    double uncertainty = config.getClockUncertainty();
    double outputDelay = config.getOutputDelay();
    double setup = library["cells"]["DFF"]["setup"];

    for (Node* n : graph.getAllNodes()) {
        n->rt = 999.0;
        
        // DFF Inputs (D)
        if (n->name.find("/D") != std::string::npos && n->name.find("reg_") != std::string::npos) {
            n->rt = period - setup - uncertainty;
        }
        
        // Output Ports
        // In Python it checked `n.name == "sum_out"`.
        // In C++ Parser, output ports might be `sum_reg_out[0]` etc. OR explicitly `sum_out` if mapped.
        // My parser logic was: if net has drivers but no loads, make it a port.
        // So `sum_reg_out[0]` would be a port.
        if (n->type == "port") {
             // Heuristic: if it's not an input port (which we set AT for), it's an output port?
             // Or check if it has incoming edges? 
             // Output ports (endpoints) usually have incoming edges and no outgoing (in this graph).
             // But DFF/D also has incoming edges.
             
             // Check if it's an output port by name or just assume all ports that are not inputs?
             // If AT is -1 (was not set as input), treat as output?
             // BUT propagation runs before RT calc. So AT should be valid.
             
             // Let's use name check "sum_" or just apply to all "port" types with valid AT?
             // Actually, apply to all designated output ports.
             // Given limitations, let's just check if it's "port" and not an input (implied by not having valid AT set by propagated inputs... wait).
             
             // Simple: If `n->edges.empty()` (no outgoing edges) and it's a port?
             if (n->edges.empty() && n->type == "port") {
                 n->rt = period - outputDelay - uncertainty;
             }
        }
    }
}

std::vector<AnalysisResult> Analysis::calculateSlack(double& worstSlack, std::string& worstNode) {
    std::cout << "Calculating Slack..." << std::endl;
    worstSlack = 999.0;
    worstNode = "";
    std::vector<AnalysisResult> results;

    for (Node* n : graph.getAllNodes()) {
        if (n->rt != 999.0 && n->at != -1.0) {
            double slack = n->rt - n->at;
            n->slack = slack;
            
            std::string status = (slack >= 0) ? "MET" : "VIOLATED";
            results.push_back({n->name, n->at, n->rt, slack, status});

            if (slack < worstSlack) {
                worstSlack = slack;
                worstNode = n->name;
            }
        }
    }
    return results;
}

File: sta_engine/analysis.py
from typing import Dict, List, Any, Tuple, Optional
from collections import deque, defaultdict
from .graph import Graph, Node

class TimingAnalyzer:
    """Performs Static Timing Analysis (STA) on the graph."""
    
    def __init__(self, graph: Graph, constraints: Dict[str, float], library: Dict[str, Any]):
        self.graph = graph
        self.constraints = constraints
        self.lib = library

    def run_analysis(self) -> Tuple[float, Optional[str], List[Dict[str, Any]]]:
        """Runs the full timing analysis pipeline."""
        self._propagate_arrival_times()
        self._calculate_required_times()
        return self._calculate_slack()

    def _propagate_arrival_times(self):
        """Propagates Arrival Times (AT) through the graph using topological traversal."""
        print("Propagating Arrival Times...")
        nodes = self.graph.get_all_nodes()
        
        # 1. Initialize and Set Start Points
        self._reset_at(nodes)
        self._apply_input_delays(nodes)
        
        # 2. Topological Sort
        topo_order = self._topological_sort(nodes)
        
        # 3. Propagate Delays
        for node in topo_order:
            if node.at == -1.0: 
                continue
                
            for target_node, delay, _ in node.edges:
                new_at = node.at + delay
                if new_at > target_node.at:
                    target_node.at = new_at

    def _reset_at(self, nodes: List[Node]):
        for node in nodes:
            node.at = -1.0

    def _apply_input_delays(self, nodes: List[Node]):
        """Sets initial arrival times for start points (Inputs, Flip-Flops)."""
        input_delay = self.constraints.get('input_delay', 0.0)
        dff_clk_q = self.lib['cells']['DFF']['delay_clk_q']

        for node in nodes:
            # Start Point: DFF Outputs (Q pin)
            if self._is_dff_output(node):
                node.at = dff_clk_q
            
            # Start Point: Primary Inputs
            # Heuristic: Known input names or simple ports
            if node.type == "port" and node.name in ["rst", "clk"]:
                 # Clock and Reset paths usually handled differently or have 0 specific delay in basic STA
                 # For data path STA, we often care about data inputs
                 pass 
            
            if "data_in" in node.name or node.name == "rst": # Heuristic from original code
                node.at = input_delay

    def _is_dff_output(self, node: Node) -> bool:
        return node.name.endswith("/Q") and "reg_" in node.name

    def _topological_sort(self, nodes: List[Node]) -> List[Node]:
        """Performs consistent topological sort for traversal."""
        in_degree = defaultdict(int)
        for node in nodes:
            for target, _, _ in node.edges:
                in_degree[target] += 1
        
        queue = deque([n for n in nodes if in_degree[n] == 0])
        topo_order = []
        
        while queue:
            node = queue.popleft()
            topo_order.append(node)
            for target, _, _ in node.edges:
                in_degree[target] -= 1
                if in_degree[target] == 0:
                    queue.append(target)
        return topo_order

    def _calculate_required_times(self):
        """Calculates Required Times (RT) based on clock period and constraints."""
        print("Calculating Required Times...")
        period = self.constraints['clock_period']
        uncertainty = self.constraints['clock_uncertainty']
        output_delay = self.constraints['output_delay']
        dff_setup = self.lib['cells']['DFF']['setup']
        
        for node in self.graph.get_all_nodes():
            # End Point: DFF Inputs (D pin)
            if self._is_dff_input(node):
                node.rt = period - dff_setup - uncertainty
            
            # End Point: Primary Outputs
            if node.name == "sum_out": # Specific to example design, should be generalized
                node.rt = period - output_delay - uncertainty

    def _is_dff_input(self, node: Node) -> bool:
         return node.name.endswith("/D") and "reg_" in node.name

    def _calculate_slack(self) -> Tuple[float, Optional[str], List[Dict[str, Any]]]:
        """Calculates slack for valid timing points."""
        print("Calculating Slack...")
        worst_slack = float('inf')
        worst_node = None
        results = []
        
        for node in self.graph.get_all_nodes():
            # Only calculate slack for constrained nodes (where RT is set)
            if node.rt == 999.0 or node.at == -1.0:
                continue
            
            slack = node.rt - node.at
            results.append({
                "node": node.name,
                "at": node.at,
                "rt": node.rt,
                "slack": slack,
                "status": "MET" if slack >= 0 else "VIOLATED"
            })
            
            if slack < worst_slack:
                worst_slack = slack
                worst_node = node.name
                    
        return worst_slack, worst_node, results
